### PR TITLE
Add IIssueLocationStore.RefreshOnBufferChanged

### LINKED ISSUE #3898: Add ILocationIssueStore.RefreshOnContentChanged method
### Description

API change required as part of handling Editor and Error list updates when suppressions change.

#### TODO
-[ ] add new IIssueLocationStore.RefreshOnContentChanged(string affectedFile) method [ does same as existing Refresh(affectedFiles), but for a single file ]
- [ ] fix up implementations
- [ ] fix tests

### PR DESCRIPTION
**Pull request by duncanp-sonar**
- fixed up existing implementations
- added tests

Fixes #3898 and #3900

### CODE REVIEW

File: src/Integration.Vsix.UnitTests/ErrorList/SonarErrorListDataSource_IssueLocationStoreTests.cs [hunk 1]
```diff
@@ -177,15 +177,53 @@ public void RefreshErrorList_FactoryIsNotRegistered_EventNotRaised()
...
+        [TestMethod]
+        public void Refresh_FactoriesWithMatchingFilesUpdatedAndErrorListRefreshed()
+            => RefreshXXX_FactoriesWithMatchingFilesUpdatedAndErrorListRefreshed(CallRefresh, shouldRaiseIssuesChanged: true);
+
+        [TestMethod]
+        public void RefreshOnBufferChanged_FactoriesWithMatchingFilesUpdatedAndErrorListRefreshed()
+            => RefreshXXX_FactoriesWithMatchingFilesUpdatedAndErrorListRefreshed(CallRefreshOnBufferChanged, shouldRaiseIssuesChanged: false);
+
+        private void RefreshXXX_FactoriesWithMatchingFilesUpdatedAndErrorListRefreshed(RefreshXXXTestOperation testOp,
+            bool shouldRaiseIssuesChanged)
```
Comment (duncanp-sonar) [inline src/Integration.Vsix.UnitTests/ErrorList/SonarErrorListDataSource_IssueLocationStoreTests.cs:226]: This is the only "common" test method that expects specific events to be raised.
I've added event checks.

None of the other tests have any functional changes to the body of the test.

File: src/Integration.Vsix.UnitTests/ErrorList/SonarErrorListDataSource_IssueLocationStoreTests.cs [hunk 2]
```diff
@@ -405,6 +480,34 @@ private void CheckSinkNotifiedOfChangeToFactories(Mock<ITableDataSink> sinkMock,
             sinkMock.Verify(x => x.FactorySnapshotChanged(It.IsAny<ITableEntriesSnapshotFactory>()), Times.Exactly(factories.Length));
         }
 
+        private static Mock<EventHandler<IssuesChangedEventArgs>> CreateAndRegisterIssuesEventHandler(SonarErrorListDataSource testSubject)
```
Comment (duncanp-sonar) [inline src/Integration.Vsix.UnitTests/ErrorList/SonarErrorListDataSource_IssueLocationStoreTests.cs:483]: A couple of new helper methods for IssueChanged event handling

File: src/IssueViz.UnitTests/Editor/LocationTagging/IssueLocationStoreAggregatorTests.cs [hunk 3]
```diff
@@ -235,6 +235,45 @@ public void Refresh_HasStores_AllStoresRefreshed()
             }
         }
 
+        [TestMethod]
+        public void RefreshOnBufferChanged_NullFilePaths_ArgumentNullException()
```
Comment (duncanp-sonar) [inline src/IssueViz.UnitTests/Editor/LocationTagging/IssueLocationStoreAggregatorTests.cs:239]: Duplicates of the existing "Refresh" tests that check the composite forwards the calls correctly.

File: src/IssueViz/Editor/LocationTagging/IssueLocationStoreAggregator.cs [hunk 4]
```diff
@@ -72,6 +72,19 @@ public IEnumerable<IAnalysisIssueLocationVisualization> GetLocations(string file
             return locationStores.SelectMany(x => x.GetLocations(filePath));
         }
 
+        public void RefreshOnBufferChanged(string affectedFilePath)
```
Comment (duncanp-sonar) [inline src/IssueViz/Editor/LocationTagging/IssueLocationStoreAggregator.cs:75]: Update the composite to forward all `RefreshOnBufferChanged` calls, same as for the existing `Refresh` implementation.

File: src/Integration.Vsix/ErrorList/SonarErrorListDataSource.cs [hunk 5]
```diff
@@ -222,6 +237,11 @@ public void Refresh(IEnumerable<string> affectedFilePaths)
                     {
                         factory.UpdateSnapshot(factory.CurrentSnapshot.GetUpdatedSnapshot());
                         InternalRefreshErrorList(factory);
+
+                        if (notifyListeners)
+                        {
+                            NotifyIssuesChanged(factory);
+                        }
```
Comment (duncanp-sonar) [inline src/Integration.Vsix/ErrorList/SonarErrorListDataSource.cs:244]: This is a simple solution that might result in duplicate notifications in rare cases involving secondary locations
e.g.
 * Factory1 has Issue1 that refers to FileA and FileB
 * Factory2 has Issue2 that refers to FileC and FileB

In that case the tagger for FileB will be notified twice.
I'll add a TODO and a separate ticket.

File: src/IssueViz/Editor/LocationTagging/IIssueLocationStore.cs [hunk 6]
```diff
@@ -37,15 +37,38 @@ public interface IIssueLocationStore
...
+        /// </summary>
+        /// <remarks>
+        /// Changing the contents of a file can affect the issues in the file:
+        /// - if content has been deleted then some issues may no longer exist.
+        /// - if content has been added or removed then the location of issues may have changed.
+        /// This method allows the service to update necessary action to update the UI.
+        /// The assumption is that Editor-related components (e.g. taggers) do not need
+        /// to updated as they will already be aware of the change.
+        /// <para>
+        /// This is a batch notification. We don't want to notify for each individual location change as that would be
```
Comment (rita-sonarsource) [inline src/IssueViz/Editor/LocationTagging/IIssueLocationStore.cs:65]: Batch comment is irrelevant

File: src/Integration.Vsix.UnitTests/ErrorList/SonarErrorListDataSource_IssueLocationStoreTests.cs [hunk 7]
```diff
@@ -177,15 +177,53 @@ public void RefreshErrorList_FactoryIsNotRegistered_EventNotRaised()
...
+            act.Should().ThrowExactly<ArgumentNullException>().And.ParamName.Should().Be("affectedFilePath");
+        }
+
+        /// <summary>
+        /// Most of the behaviour of Refresh(file) and RefreshOnBufferChanged(files) is the same.
+        /// This delegate is used by tests that check the common functionality
+        /// i.e. there is a common internal test method that does the common setup and assertions.
+        /// This delegate is passed into that common method to make a call to either 
+        /// "Refresh(file)" or "RefreshOnBufferChanged(files)"
+        /// </summary>
```
[the remark this answers is not in the document]
Comment (rita-sonarsource) [reply src/Integration.Vsix.UnitTests/ErrorList/SonarErrorListDataSource_IssueLocationStoreTests.cs:201]: Hmm. It's not bad IMO. Or maybe it's already so bad that I can't see it...

Also maybe worth renaming XXX to Impl, as in, RefreshImpl, or RefreshInner

**Review by rita-sonarsource — APPROVED** [review]

File: src/Integration.Vsix.UnitTests/ErrorList/SonarErrorListDataSource_IssueLocationStoreTests.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (duncanp-sonar) [reply src/Integration.Vsix.UnitTests/ErrorList/SonarErrorListDataSource_IssueLocationStoreTests.cs:201]: I think it's more an indicator that there might be a design problem. Still, not a problem for today.

**Pull request merged**